### ISSUE TITLE
JSON form submission

### ISSUE THREAD

**Issue by FlorianRappl** [AngleSharp/AngleSharp#126]
An upcoming addition to the possible form encoding types is `application/json`. The idea is to submit a JSON string. This string can also contain files (base64 encoded). An [unofficial draft](https://darobin.github.io/formic/specs/json/) exists.

The idea is to include this for form submission. The implementation should be fairly straight forward. This is auxiliary and therefore won't have very high priority. But it would be still nice to have.

**Comment by Suchiman** [comment]
I'm working on it and i've implemented most of the described algorithm, but i stumpled upon the json infrastructure. Should JSONObject / JSONArray and Stringify be reimplemented for this?

**Comment by FlorianRappl** [comment]
That is a good question. My initial thought was to not use any of those, but to build a JSON string immediately. Of course things can get messy and hence you need some intermediate object representation to build that string. I would go for `Dictionary<String, JsonElement>`, where `JsonElement` is an abstract base class with derived classes such as `JsonArray`, `JsonObject`, ...

These classes are only for book-keeping and they have one thing in common - they override `ToString()`. Otherwise they have no fancy API or anything. They are just POD and should stay internal I would say.

A different approach would be to have a JSON service that defines a rich API. But defining that API is possibly already most of the work. The form submission would simply call for that API and (if available) use the resulting methods to create the JSON object. The API would then also be used to stringify the object.

What are your thoughts?

**Comment by Suchiman** [comment]
The former approach would probably work, the latter would be interesting when it is combined with AngleSharp.Scripting since Scripting / Jint probably has all required infrastructure, but since this isn't a required dependency, it wouldn't work without it.

Another thing i need to figure out, is if i can keep the current structure of how `FormDataSet` calls each `FormDataSetEntry` to serialize itself.

**Comment by FlorianRappl** [comment]
Yes, that will be interesting. I am not really satisfied with the current solution, but a different path looked like over-engineering from a first glance.

If there are problems we need to rethink that! Just tell me about your progress.

Thanks for your efforts!

**Comment by FlorianRappl** [comment]
What's the current status on this? Is there anything I can help you with?

**Comment by Suchiman** [comment]
Hi, sorry i was slacking in the last two weeks.
I've put a standalone working prototype up here https://github.com/Suchiman/JsonFormPrototype
It's not optimized yet but strictly conforms to the spec and produces the same output for all samples.

**TODOS**:
- [x] Decouple from Newtonsoft (used it for now to not having to worry about JSON serializing itself)
- [x] Integration in AngleSharp (prototype operates on a copy of FormDataSet but exposes the private properties)

**Comment by FlorianRappl** [comment]
Alright - no problem (I was just curious). I will try to play around with your prototype - maybe I'll find a good way to get rid of any JSON dependencies directly / externalize it without much trouble.

Also I'll try to alleviate / inspect the second point.

**Comment by Suchiman** [comment]
Current status: I've written a naive json implementation and figured how to integrate it into AngleSharp, i think i can make an integrated prototype over the weekend.
Further TODOs:
- [ ] Json implementation needs a bit of hardening like correct escaping
- [ ] Investigate test case 2, input type numbers must be converted to non quoted integers apparently

**Comment by FlorianRappl** [comment]
That is great news!
The JSON implementation you provide can also be very lightweight. If it is exchangeable, then anyone can always just use their favorite (or own) JSON parser. So I wouldn't worry too much about 1).

One of the remaining questions is if we should include this directly in the core, or if the core should expose the possibility of being extended with new methods for form submission. The latter would be my favorite, although I would also like to have it in the core by default.

**Comment by Suchiman** [comment]
Yeah my JSON implementation (which doesn't parse but only outputs) is as lightweight as it could be, barely sufficient to support this scenario.

Also i don't see much benefit into making this implementation detail exchangeable. It's nothing the user itself will / should ever touch and therefore does not need to worry about API surface preference, nor should he care about it, assuming it generates correct json.

Neither does creating a 10kb dll with this small functionality seem sensible for me.
Allowing the user to add custom (non standard, assuming all standards are supported) methods of form submission seems also questionable.
Assuming form submission support is used to emulate browsers, it does not need support anything more than them and with working Scripting, they can write a custom request using JS or, when specially extending AngleSharp, using `<script type="text/c-sharp">` :smile: 

I've successfully integrated the algorithm into `FormDataSet.cs` but i've stumpled upon something unclear. Submitting JSON clearly needs a POST body but the spec never mentions a form submission method and none of the samples specify any method. Is `method="POST"` implicit then when `enctype="application/json"` ?

Should i remove spec comments from the final code?

Also what is your opinion on C#6 compiler features (none which require framework additions) ?

**Comment by FlorianRappl** [comment]
> Is method="POST" implicit then when enctype="application/json" ?

I assume that if `method=POST` the generated JSON string will be transported in the body and that if `method=GET` the JSON string will be URL encoded and transported in the query part of the URL. But you are right, the specification is not very specific on this topic.

> Should i remove spec comments from the final code?

I would leave the spec. comments in the code. I always find it interesting to have links back to the outlined algorithm. In AngleSharp I often placed the basic set of information in the documentation comment for, e.g., a class, or a method.

> Also what is your opinion on C#6 compiler features (none which require framework additions) ?

At some point I want to convert AngleSharp to C# 6 (mostly using the Elvis `?.` to replace explicit null-checks, but also some readonly properties and initializers). However, I fear that the current adaption rate is unfortunately too low to convert the project right now. I would therefore wait until next year before introducing C# 6 features to the code.

Regarding the integration and extension you are right: There is always the ability to use the `submit` event for custom submission implementations.

In summary: Looks all good, I would (for the moment) change C# 6 to 5, but the rest certainly looks fine.

**Comment by Suchiman** [comment]
Another thing i've come upon are apparently unspecified conversions.
E.g. in Example1, `"on"` from checkbox is converted to `true` and in Example2, the value is converted to a number.
I could add code to convert to decimal when TextDataSetEntry.Type == "number" and when Type == "checkbox" then convert on to true but that's all rather guessing.

Anyway i've committed the integrated prototype here https://github.com/Suchiman/AngleSharp/commit/85195545eab411ddcfa77ec2a49d6343e88f728a and it passes examples 3-10

**Comment by FlorianRappl** [comment]
Alright I placed the changes in a new branch. I will look at it later - but seems to be very good so far.
https://github.com/AngleSharp/AngleSharp/tree/feature-%23126

The `application/json` form submission type will be live with v0.9.2.

**Comment by Suchiman** [comment]
For correct JSON string encoding i've glimpsed at https://github.com/Microsoft/referencesource/blob/74706335e3b8c806f44fa0683dc1e18d3ed747c2/System.Web/Util/HttpEncoder.cs#L391-L459 which lives in System.Web but AngleSharp is on Portable.
You could copy it but i'm unsure about license.
